README.md
# IQOption Trading Bot 

## Warning: 

  The bots were coded using an API from IQ Option platform. Based on my experience, this is an extremely bad broker with very high spread and low fps. As this platform was most accessible from where I live, I decided to use it to test my trading strategies. The documentation of the APIs were also limited and lackluster. So I would recommend staying away from this API and use other reputable exchanges. That being said, we will strictly discuss the strategies used and tested, instead of the installing and running applications.  

## Introduction 

This spring of 2021 was my first gap semester and I found a new interest: investing and trading. As a coder, I wanted to automate my trading experience as well as improving my strategies using bots. This is a great way to remove emotions out of my decisions and induce a disciplinary framework for my trading strategy. An effective and consistent system would ultimately become a good source of passing income. 

## Strategies: 

1. Momentum catch: 

  The bot will detect momentum shift in a specific time frame and place order following that momentum. 
  
  For examplem, if Bitcoin shows 3 green consecutive 5m candles, then the bot will put a call order. A sell order will be placed in the opposite scenarios. 
  
  Pros: good for a bear trend or bull trend 
  
  Cons: bad during a sideway or crab market 
  
  Please check btc5m.py, btc1m.py, stock1m.py for references.


2. dca bot: 

  The bot will continue to buy as the price drops to dollar cost average the call order spanning through out a dip. Then it will sell all positions once the price recovers to a certain point.
  
  Pros: works very well during volatile sideway movements 
  
  Cons: works very poorly during trendy market(bear or bull) 
  

## results: 

- The momentum bots' success rate were about 50%, there are still many tweaking that needs to be done and implementation of different indicators for better decision making. 

- The dca bot works really well at first, with a success rate of almost 100%. However, during the Bitcoin May crash, it fails miserably due to the price not recovering. If u run this bot for years then it would still work as Bitcoin recovers later in a Bull market. But as trading is all about short term movements, this method renders failure in an event of a big bear crash. 

## Google cloud implementaion 

- The bots were tested 24/7 using Google Cloud Computing Machines 

- Tools include: 

1. python3 
2. dependencies
3. tmux to run the script 24/7, please refer to tmux documentation for usage 
4. vim as the virtual IDE. 


## Conclusion 

As I will be migrating my trading bot development to different platform and a different API, This serves as a small introduction only. Hence, I did not discuss too much into the coding aspect. Please look at my code yourself or ask me any questions if you wanna learn more about any specific lines or functions. 

Trading bot algorithms are extremely hard to build and test. I will be working on during my passive time in college to find the most optimized strategy. During this period I will be uploading a more thorough documentation and testings as well as a better code representation. 


history.txt

Volswagen
position: buy
entry: 211.07500000000002
close: 210.75
profit: -0.77%


Volswagen
position: sell
entry: 210.225
close: 209.70000000000002
profit: 1.25%


Volswagen
position: sell
entry: 209.375
close: 209.9
profit: -1.25%


Volswagen
position: buy
entry: 210.20000000000002
close: 209.89999999999998
profit: -0.71%


Volswagen
position: buy
entry: 210.55
close: 210.25
profit: -0.71%


Volswagen
position: sell
entry: 209.67499999999998
close: 210.025
profit: -0.83%


Volswagen
position: sell
entry: 209.725
close: 209.275
profit: 1.07%


Volswagen
position: sell
entry: 208.875
close: 209.175
profit: -0.72%


Volswagen
position: sell
entry: 209.02499999999998
close: 209.375
profit: -0.84%


Volswagen
position: buy
entry: 209.875
close: 209.725
profit: -0.36%


Volswagen
position: sell
entry: 209.52499999999998
close: 209.675
profit: -0.36%


Volswagen
position: buy
entry: 209.85000000000002
close: 210.075
profit: 0.54%


Volswagen
position: sell
entry: 210.0
close: 210.3
profit: -0.71%


Volswagen
position: sell
entry: 210.0
close: 210.20000000000002
profit: -0.48%


Volswagen
position: sell
entry: 209.79999999999998
close: 210.05
profit: -0.60%


Volswagen
position: buy
entry: 210.15
close: 209.85
profit: -0.71%


BTC
position: sell
entry: 58753.836
close: 58949.727
profit: -1.00%


BTC
position: sell
entry: 58760.566
close: 58964.996
profit: -1.04%


BTC
position: buy
entry: 58872.113
close: 58656.066
profit: -1.10%


BTC
position: sell
entry: 58560.055
close: 58766.984
profit: -1.06%


BTC
position: sell
entry: 58558.836
close: 58708.273
profit: -0.77%


BTC
position: sell
entry: 58404.445
close: 58536.363
profit: -0.68%



test.py

a = 5
def change(a):
    a = a*2

print(a)

btc2m.py
import time
from iqoptionapi.stable_api import IQ_Option
import time

def compare(prev, curr):
    if curr>prev:
        return "high"
    elif curr < prev: 
        return "low"
    else:
        return 0

history = []
order = False 
mybot=IQ_Option("acao2020@macduffie.org",input())
mybot.connect()#connect to iqoption

mybot.change_balance("PRACTICE")
#order,id1 = mybot.buy_order("crypto", "BTCUSD","buy",1, 2,"market")

opcode_data=mybot.get_all_ACTIVES_OPCODE()

#invest = float(mybot.get_async_order(id1)["position-changed"]["msg"]['invest_enrolled'])
mybot.subscribe_top_assets_updated("crypto")

def opcode_to_name(opcode_data,opcode):
    return list(opcode_data.keys())[list(opcode_data.values()).index(opcode)] 

previous_price = 0
count = 0
id1 = 0
invest = 0
position =""

def closePosition(id,percent, position, entry, close):

    if percent > 2.5 or percent < -2:
        mybot.close_position(id)
        print("closed " + str(percent))
        f = open("btc2m.txt", "a")
        f.write("\n")
        f.write("position: "+ str(position))
        f.write("\n")
        f.write("entry: "+ str(entry))
        f.write("\n")
        f.write("close: "+ str(close))
        f.write("\n")
        percent = ('%.2f' %percent)
        f.write("profit: " + str(percent)+"%")
        f.write("\n")
        f.write("\n")
        f.close()
        return False
    else:
        return True


while True:
    assets = mybot.get_top_assets_updated("crypto")
    current_price = 0
    if assets:
        BTC = assets[0]
        current_price = BTC["cur_price"]["value"]

    if not order and previous_price != 0:
        #candle manipulation
        if compare(previous_price,current_price) =="high" and count>=0:
            print("green")
            count +=1
            previous_price = current_price
        elif compare(previous_price,current_price)=="high" and count< 0:
            print("green")
            count = 1
            previous_price = current_price
        elif compare(previous_price,current_price)=="low" and count <=0:
            print("red")
            count-=1
            previous_price = current_price
        elif compare(previous_price,current_price)=="low" and count > 0:
            print("red")
            count = -1
            previous_price = current_price
        
        if count == -3:
            count = 0
            print("sell")
        
            order1,id1 = mybot.buy_order("crypto", "BTCUSD","sell",100, 2,"market",None, None,)
            invest = float(current_price) - 77
            position = "sell"
            order = True


        elif count == 3:
            count = 0
            print("buy")
            order1,id1 = mybot.buy_order("crypto", "BTCUSD","buy",100, 2,"market",None, None,) 
            invest = float(current_price) + 77
            position = "buy"
            order = True

        time.sleep(120)
        
    elif order:
        if position == "sell":
                current = float(current_price) + 77
                print(position +": "+ str(invest))
                print("current: " + str(current))
                profit = -(current - invest)/invest*100 *2
                print("profit: "+ ('%.2f' %profit) + "%")
                order = closePosition(id1,profit,position,invest,current)
        else:

                current = float(current_price) - 77
                print(position +": "+ str(invest))
                print("current: " + str(current))
                profit = (current - invest)/invest*100*2
                print("profit: "+ ('%.2f' %profit) + "%")
                order = closePosition(id1,profit,position,invest,current)

        previous_price = current_price
        time.sleep(30)


    else:
        time.sleep(30)
        previous_price = current_price


    


print(mybot.get_balance())

btc3m.py
import time
from iqoptionapi.stable_api import IQ_Option
import time

def compare(prev, curr):
    if curr>prev:
        return "high"
    elif curr < prev: 
        return "low"
    else:
        return 0

print("username: ")
username = input()
print("password: ")
password = input()

history = []

order = False 
mybot=IQ_Option(username,password)

mybot.connect()#connect to iqoption

mybot.change_balance("PRACTICE")
#order,id1 = mybot.buy_order("crypto", "BTCUSD","buy",1, 2,"market")

opcode_data=mybot.get_all_ACTIVES_OPCODE()

#invest = float(mybot.get_async_order(id1)["position-changed"]["msg"]['invest_enrolled'])
mybot.subscribe_top_assets_updated("crypto")

def opcode_to_name(opcode_data,opcode):
    return list(opcode_data.keys())[list(opcode_data.values()).index(opcode)] 

previous_price = 0
count = 0
id1 = 0
invest = 0
position =""
ath = 0
profit_take = 2
stop_loss = -1.5
profit_margin = 0.2

win = 0 
loss = 0

def closePosition(id,percent, position, entry, close, ath, profit_take, profit_margin, stop_loss):
    if percent < stop_loss or (ath!= 0 and (ath - percent >= profit_margin)):
        global win 
        global loss
        if percent < stop_loss:
            loss+=1
        else:
            win+=1
        mybot.close_position(id)
        print("closed " + str(percent))
        f = open("btc3m.txt", "a")
        f.write("\n")
        f.write("BTC")
        f.write("\n")
        f.write("position: "+ str(position))
        f.write("\n")
        f.write("entry: "+ str(entry))
        f.write("\n")
        f.write("close: "+ str(close))
        f.write("\n")
        percent = ('%.2f' %percent)
        f.write("profit: " + str(percent)+"%")
        f.write("\n")
        f.write("win: " + str(win)+ " loss: "+ str(loss))
        f.write("\n")
        f.write("\n")
        f.close()
        ath = 0
        return False,ath

    elif percent > profit_take:
        print("ye")
        print(percent)
        print(ath)
        if percent > ath:
            print("athh")
            ath = percent
        return True,ath
    else:
        return True,ath


while True:
    assets = mybot.get_top_assets_updated("crypto")
    current_price = 0
    if assets:
        V = assets[0]
        current_price = V["cur_price"]["value"]

    if not order and previous_price != 0:
        #candle manipulation
        if compare(previous_price,current_price) =="high" and count>=0:
            print("green")
            count +=1
            previous_price = current_price
        elif compare(previous_price,current_price)=="high" and count< 0:
            print("green")
            count = 1
            previous_price = current_price
        elif compare(previous_price,current_price)=="low" and count <=0:
            print("red")
            count-=1
            previous_price = current_price
        elif compare(previous_price,current_price)=="low" and count > 0:
            print("red")
            count = -1
            previous_price = current_price
        
        if count == -3:
            count = 0
            print("sell")
        
            order1,id1 = mybot.buy_order("crypto", "BTCUSD","sell",100, 2,"market",None, None,)
            invest = float(mybot.get_position(id1)[1]["position"]["open_underlying_price"])
            position = "sell"
            order = True


        elif count == 3:
            count = 0
            print("buy")
            order1,id1 = mybot.buy_order("crypto", "BTCUSD","buy",100, 2,"market",None, None,)
            invest = float(mybot.get_position(id1)[1]["position"]["open_underlying_price"])
            position = "buy"
            order = True

        time.sleep(180)
        
    elif order:
        if position == "sell":
                current = float(current_price) + 77
                print(position +": "+ str(invest))
                print("current: " + str(current))
                profit = -(current - invest)/invest*100 *2
                print("profit: "+ ('%.2f' %profit) + "%")
                print("ath: " + str(ath))
                order,ath = closePosition(id1,profit,position,invest,current,ath,profit_take,profit_margin,stop_loss)
        else:

                current = float(current_price) - 77
                print(position +": "+ str(invest))
                print("current: " + str(current))
                profit = (current - invest)/invest*100*2
                print("profit: "+ ('%.2f' %profit) + "%")
                print("ath: " + str(ath))
                order,ath = closePosition(id1,profit,position,invest,current,ath,profit_take,profit_margin,stop_loss)

        previous_price = current_price
        time.sleep(10)


    else:
        time.sleep(10)
        previous_price = current_price


    




dcabotbtc.py
import time
from iqoptionapi.stable_api import IQ_Option
import time

def compare(prev, curr):
    if curr>prev:
        return "high"
    elif curr < prev: 
        return "low"
    else:
        return 0

history = []
order = False 
mybot=IQ_Option(input("username:"),input("password"))
mybot.connect()#connect to iqoption

mybot.change_balance("PRACTICE")
#order,id1 = mybot.buy_order("crypto", "BTCUSD","buy",1, 2,"market")

opcode_data=mybot.get_all_ACTIVES_OPCODE()

#invest = float(mybot.get_async_order(id1)["position-changed"]["msg"]['invest_enrolled'])
mybot.subscribe_top_assets_updated("crypto")

def opcode_to_name(opcode_data,opcode):
    return list(opcode_data.keys())[list(opcode_data.values()).index(opcode)] 


def percent_change(compare_price, current_price):
    change = (current_price - compare_price)/compare_price*100 *2
    change = '%.2f'%change
    return float(change)

def reset():
    for id in positions:
        mybot.close_position(id)


    order1,id1 = mybot.buy_order("crypto", "BTCUSD","buy",50, 2,"market",None, None,)
    positions.append(id1)
    compare_price = float(mybot.get_position(id1)[1]["position"]["open_underlying_price"])



positions =[]
order1,id1 = mybot.buy_order("crypto", "BTCUSD","buy",50, 2,"market",None, None,)

positions.append(id1)
compare_price = float(mybot.get_position(id1)[1]["position"]["open_underlying_price"])
current_price = compare_price
invest = 50
orders = 1
win = False
win_percent = 0

lossdca = -2
profit_margin = 0.5
profit = 2
dca_factor = 1.5

f = open("dcav2.txt", "a")
f.write("\n")
f.write("buy " + str(invest) + " at " +str(current_price))
f.write("\n")
f.close()

while True:
    assets = mybot.get_top_assets_updated("crypto")
    if assets:
        BTC = assets[0]
        current_price = BTC["cur_price"]["value"]

    print("BTC: ")
    print("compare price: "+ str(compare_price))
    print("current price: "+ str(current_price))
    percent = percent_change(compare_price,current_price)
    print(str(percent)+ " %")
    
    if percent < lossdca and not win and orders < 6:
        orders += 1
        invest = invest*dca_factor
        compare_price = current_price
        lossdca -= 0.5
        profit + 0.7
        order1,id1 = mybot.buy_order("crypto", "BTCUSD","buy",invest, 2,"market",None, None,)
        f = open("dcav2.txt", "a")
        f.write("\n")
        f.write("buy " + str(invest) + " at " +str(current_price))
        f.write("\n")
        f.close()
        positions.append(id1)
    
    elif percent > profit and not win:
        win_percent = percent
        win = True
    
    elif win:
        print("ath: "+ str(win_percent))
        if compare(win_percent, percent) == "high":
            win_percent = percent
        elif (win_percent - percent) >= profit_margin:
            win = False
            f = open("dcav2.txt", "a")
            f.write("\n")
            f.write("close order")
            f.write("\n")
            f.write("profit from lowest trade: "+ str(percent))
            f.write("\n")
            f.close()
            reset()
            invest = 50
            orders = 1
            lossdca = -2
            profit = 2
            compare_price = current_price
    
    
    time.sleep(30)
    
    

stock4mtrend.py
import time
from iqoptionapi.stable_api import IQ_Option
import time

def compare(prev, curr):
    if curr>prev:
        return "high"
    elif curr < prev: 
        return "low"
    else:
        return 0

print("username: ")
username = input()
print("password: ")
password = input()

history = []

order = False 
mybot=IQ_Option(username,password)

mybot.connect()#connect to iqoption

mybot.change_balance("PRACTICE")
#order,id1 = mybot.buy_order("crypto", "BTCUSD","buy",1, 2,"market")

opcode_data=mybot.get_all_ACTIVES_OPCODE()

#invest = float(mybot.get_async_order(id1)["position-changed"]["msg"]['invest_enrolled'])
mybot.subscribe_top_assets_updated("cfd")

def opcode_to_name(opcode_data,opcode):
    return list(opcode_data.keys())[list(opcode_data.values()).index(opcode)] 

previous_price = 0
count = 0
id1 = 0
invest = 0
position =""
ath = 0
profit_take = 3
stop_loss = -1.5
profit_margin = 0.4

def closePosition(id,percent, position, entry, close, ath, profit_take, profit_margin, stop_loss):

    if percent < stop_loss or (ath!= 0 and (ath - percent >= profit_margin)):
        mybot.close_position(id)
        print("closed " + str(percent))
        f = open("stock4m.txt", "a")
        f.write("\n")
        f.write("Volswagen")
        f.write("\n")
        f.write("position: "+ str(position))
        f.write("\n")
        f.write("entry: "+ str(entry))
        f.write("\n")
        f.write("close: "+ str(close))
        f.write("\n")
        percent = ('%.2f' %percent)
        f.write("profit: " + str(percent)+"%")
        f.write("\n")
        f.write("\n")
        f.close()
        ath = 0
        return False,ath

    if percent > profit_take:
        if percent > ath:
            ath = percent
        return True,ath
    else:
        return True,ath


while True:
    assets = mybot.get_top_assets_updated("cfd")
    current_price = 0
    if assets:
        V = assets[304]
        current_price = V["cur_price"]["value"]

    if not order and previous_price != 0:
        #candle manipulation
        if compare(previous_price,current_price) =="high" and count>=0:
            print("green")
            count +=1
            previous_price = current_price
        elif compare(previous_price,current_price)=="high" and count< 0:
            print("green")
            count = 1
            previous_price = current_price
        elif compare(previous_price,current_price)=="low" and count <=0:
            print("red")
            count-=1
            previous_price = current_price
        elif compare(previous_price,current_price)=="low" and count > 0:
            print("red")
            count = -1
            previous_price = current_price
        
        if count == -2:
            count = 0
            print("sell")
        
            order1,id1 = mybot.buy_order("cfd", "VOW3D-CHIX","sell",100, 5,"market",None, None,)
            invest = float(current_price) - 0.05
            position = "sell"
            order = True


        elif count == 2:
            count = 0
            print("buy")
            order1,id1 = mybot.buy_order("cfd", "VOW3D-CHIX","buy",100, 5,"market",None, None,)
            invest = float(current_price) + 0.05
            position = "buy"
            order = True

        time.sleep(300)
        
    elif order:
        if position == "sell":
                current = float(current_price) + 0.05
                print(position +": "+ str(invest))
                print("current: " + str(current))
                profit = -(current - invest)/invest*100 *5
                print("profit: "+ ('%.2f' %profit) + "%")
                print("ath: " + str(ath))
                order,ath = closePosition(id1,profit,position,invest,current,ath,profit_take,profit_margin,stop_loss)
        else:

                current = float(current_price) - 0.05
                print(position +": "+ str(invest))
                print("current: " + str(current))
                profit = (current - invest)/invest*100*5
                print("profit: "+ ('%.2f' %profit) + "%")
                print("ath: " + str(ath))
                order,ath = closePosition(id1,profit,position,invest,current,ath,profit_take,profit_margin,stop_loss)

        previous_price = current_price
        time.sleep(15)


    else:
        time.sleep(15)
        previous_price = current_price


    




stock1m.py
import time
from iqoptionapi.stable_api import IQ_Option
import time

def compare(prev, curr):
    if curr>prev:
        return "high"
    elif curr < prev: 
        return "low"
    else:
        return 0

print("username: ")
username = input()
print("password: ")
password = input()

history = []

order = False 
mybot=IQ_Option(username,password)

mybot.connect()#connect to iqoption

mybot.change_balance("PRACTICE")
#order,id1 = mybot.buy_order("crypto", "BTCUSD","buy",1, 2,"market")

opcode_data=mybot.get_all_ACTIVES_OPCODE()

#invest = float(mybot.get_async_order(id1)["position-changed"]["msg"]['invest_enrolled'])
mybot.subscribe_top_assets_updated("cfd")

def opcode_to_name(opcode_data,opcode):
    return list(opcode_data.keys())[list(opcode_data.values()).index(opcode)] 

previous_price = 0
count = 0
id1 = 0
invest = 0
position =""
ath = 0
profit_take = 1.5
stop_loss = -1
profit_margin = 0.2
win = 0 
loss = 0

def closePosition(id,percent, position, entry, close, ath, profit_take, profit_margin, stop_loss):
    global win
    global loss
    if percent < stop_loss or (ath!= 0 and (ath - percent >= profit_margin)):
        if percent < stop_loss:
            loss+=1
        else:
            win+=1

        mybot.close_position(id)
        print("closed " + str(percent))
        f = open("stock1m.txt", "a")
        f.write("\n")
        f.write("Volswagen")
        f.write("\n")
        f.write("position: "+ str(position))
        f.write("\n")
        f.write("entry: "+ str(entry))
        f.write("\n")
        f.write("close: "+ str(close))
        f.write("\n")
        percent = ('%.2f' %percent)
        f.write("profit: " + str(percent)+"%")
        f.write("\n")
        f.write("win: " + str(win)+ " loss: "+ str(loss))
        f.write("\n")
        f.write("\n")
        f.close()
        ath = 0
        return False,ath

    if percent > profit_take:
        if percent > ath:
            ath = percent
        return True,ath
    else:
        return True,ath


while True:
    assets = mybot.get_top_assets_updated("cfd")
    current_price = 0
    if assets:
        V = assets[304]
        current_price = V["cur_price"]["value"]

    if not order and previous_price != 0:
        #candle manipulation
        if compare(previous_price,current_price) =="high" and count>=0:
            print("green")
            count +=1
            previous_price = current_price
        elif compare(previous_price,current_price)=="high" and count< 0:
            print("green")
            count = 1
            previous_price = current_price
        elif compare(previous_price,current_price)=="low" and count <=0:
            print("red")
            count-=1
            previous_price = current_price
        elif compare(previous_price,current_price)=="low" and count > 0:
            print("red")
            count = -1
            previous_price = current_price
        
        if count == -3:
            count = 0
            print("sell")
        
            order1,id1 = mybot.buy_order("cfd", "VOW3D-CHIX","sell",100, 5,"market",None, None,)
            invest = float(mybot.get_position(id1)[1]["position"]["open_underlying_price"])-0.05
            position = "sell"
            order = True


        elif count == 3:
            count = 0
            print("buy")
            order1,id1 = mybot.buy_order("cfd", "VOW3D-CHIX","buy",100, 5,"market",None, None,)
            invest = float(mybot.get_position(id1)[1]["position"]["open_underlying_price"])+0.05
            position = "buy"
            order = True

        time.sleep(60)
        
    elif order:
        if position == "sell":
                current = float(current_price) + 0.05
                print(position +": "+ str(invest))
                print("current: " + str(current))
                profit = -(current - invest)/invest*100 *5
                print("profit: "+ ('%.2f' %profit) + "%")
                print("ath: " + str(ath))
                order,ath = closePosition(id1,profit,position,invest,current,ath,profit_take,profit_margin,stop_loss)
        else:

                current = float(current_price) - 0.05
                print(position +": "+ str(invest))
                print("current: " + str(current))
                profit = (current - invest)/invest*100*5
                print("profit: "+ ('%.2f' %profit) + "%")
                print("ath: " + str(ath))
                order,ath = closePosition(id1,profit,position,invest,current,ath,profit_take,profit_margin,stop_loss)

        previous_price = current_price
        time.sleep(30)


    else:
        time.sleep(30)
        previous_price = current_price


    


